### DISCUSSION TITLE
Setting a cursor for .NET MAUI VisualElement

### DISCUSSION THREAD

**Discussion by giscus** [VladislavAntonyuk/vladislavantonyuk.github.io#98] [Announcements]
# Setting a cursor for .NET MAUI VisualElement

The article demonstrates how to change cursor of any .NET MAUI VisualElement.

https://vladislavantonyuk.github.io/articles/Setting-a-cursor-for-.NET-MAUI-VisualElement/

**Comment by jerry91122** [comment]
how to trigger actions when the mouse is clicked?

**Reply by VladislavAntonyuk** [reply]
Add GestureRecognizer to your element

**Comment by skane526** [comment]
Hello, where do I put this code? MyVisualElement.SetCustomCursor(CursorIcon.Hand, MyVisualElement.Handler?.MauiContext);

Should this be in my MainPage.xaml.cs file?

Thanks!

**Reply by VladislavAntonyuk** [reply]
Yes, you put it in place where you want to set a cursor

**Comment by sreckcep** [comment]
Hello, I like your tutorial, but I have a problem with getting the MauiContext. I don't use AppShell and when I directly put mainPage = HomePage() (which is just a contentPage) in app.xaml.cs, I get an exception because the mauicontext is null when building the view from xaml. This didn't happen with appshell. Any idea ?

**Reply by VladislavAntonyuk** [reply]
MauiContext usually sets when the page is loaded, just to try to set the cursor a bit later. I can take a look if you provide a reproduction sample

**Reply by sreckcep** [reply]
I tried setting the cursors in the code-behind in OnHandledChanged(), and it worked, thank you. However that seems to mean I cannot use the convenient xaml syntax, which is a shame.

**Reply by VladislavAntonyuk** [reply]
you can play with the MauiContext. In attachedProperty, I used `visualElement.SetCustomCursor((CursorIcon)newvalue, Application.Current?.MainPage?.Handler?.MauiContext);`, that may be null in some cases. Set the value that is relevant for your app

**Reply by sreckcep** [reply]
I did try to fiddle with that, but I just couldn't get a MauiContext at the xaml parsing time. It looks like there is just no context available until my first OnHandlerChanged is called, which is after the page is constructed from xaml... be it `Application.Current.Handler` or `Application.Current.MainPage.Handler` or any Handler really. Maybe there's a hidden Handler somewhere that is initialized at that point but I couldn't find it.

**Reply by sreckcep** [reply]
by the way, thank you for taking the time to help me. That means a lot, especially since you're pretty much the only meaningful source about this topic online ! :)

**Reply by VladislavAntonyuk** [reply]
I will try later with the page (without Shell). Am I right, you have in App.cs: MainPage = new YourContentPage()?

**Reply by sreckcep** [reply]
Indeed.

**Reply by sreckcep** [reply]
Here is my App.cs page. I can't disclose my page code because it is company code however

```cs
        public VeritoolApp()
        {
            string lang = Thread.CurrentThread.CurrentCulture.TwoLetterISOLanguageName;
            Thread.CurrentThread.CurrentUICulture = new System.Globalization.CultureInfo(lang);
            InitializeComponent();
            




#if WINDOWS
            Microsoft.Maui.Handlers.WindowHandler.Mapper.AppendToMapping(nameof(IWindow), (handler, view) =>
            {

                var mauiWindow = handler.VirtualView;
                var nativeWindow = handler.PlatformView;
                nativeWindow.Activate();
                IntPtr windowHandle = WinRT.Interop.WindowNative.GetWindowHandle(nativeWindow);
                WindowId windowId = Microsoft.UI.Win32Interop.GetWindowIdFromWindow(windowHandle);
                AppWindow appWindow = Microsoft.UI.Windowing.AppWindow.GetFromWindowId(windowId);
                
                if (appWindow.Presenter is OverlappedPresenter p) 
                {
                    p.IsResizable = true;
                    p.Maximize();
                    
                }
                //appWindow.Resize(new SizeInt32(WindowWidth, WindowHeight));
            });
#endif

            MainPage = new HomePage(); 
        }
```

**Reply by ghost** [reply]
For anyone facing `null` context issue, I was setting the cursor for controls before the main page was created, essentially leading to `null` MAUI context. I've resolved it by monitoring, when the context gets changed. The updated `CursorBehavior` worked as a charm, for me.

```csharp
    public class CursorBehavior
    {
        public static readonly BindableProperty CursorProperty = BindableProperty.CreateAttached("Cursor", typeof(CursorIcon), typeof(CursorBehavior), CursorIcon.Arrow, propertyChanged: CursorChanged);

        private static void CursorChanged(BindableObject bindable, object oldvalue, object newvalue)
        {
            if (bindable is VisualElement visualElement)
            {
                visualElement.HandlerChanged += (bindable, e) => {
                    visualElement.SetCustomCursor((CursorIcon)newvalue, Application.Current?.MainPage?.Handler?.MauiContext);
                };
            }
        }

        public static CursorIcon GetCursor(BindableObject view) => (CursorIcon)view.GetValue(CursorProperty);

        public static void SetCursor(BindableObject view, CursorIcon value) => view.SetValue(CursorProperty, value);
    }
```

**Comment by Saidalip1** [comment]
Is it possible to add our own cursor when hovering over a control? We have some core cursors that are working. Is it possible to add a new .cur file on Windows and Mac in .NET MAUI?

**Comment by tekmun** [comment]
In maui iOS .Net 9, this code throws an error:

var path = UIBezierPath.FromPath(pathGeometry.ToCGPath().Data);

I think using Microsoft.Maui.Controls.Compatibility.Platform.iOS; is no longer valid in .Net 9

Please kindly help. Thanks.

**Reply by VladislavAntonyuk** [reply]
See updated code in repo.

**Comment by tekmun** [comment]
The suggested solution does not work.
I removed the if-else condition and only use the if condition result.
path = UIBezierPath.FromPath(new CGPath());
